core/domain/html_cleaner.py

```diff
@@ -782,11 +783,202 @@ def add_caption_attr_to_image(html_string):
+
+        for collapsible in soup.findAll(
+                name='oppia-noninteractive-collapsible'):
+            # This is wrapped in try catch block since any invalid
+            # content such as int will not be convertible in Beautiful
+            # Soup format.
+            try:
+                content_html = json.loads(
+                    unescape_html(collapsible['content-with-value']))
+                soup_for_collapsible = bs4.BeautifulSoup(
+                    content_html, 'html.parser')
+                _validate_customization_args_in_soup(
+                    soup_for_collapsible, err_dict, html_string)
+            except Exception:
+                if 'oppia-noninteractive-collapsible' not in err_dict:
+                    err_dict['oppia-noninteractive-collapsible'] = html_string
+                elif html_string not in err_dict[
+                        'oppia-noninteractive-collapsible']:
+                    err_dict['oppia-noninteractive-collapsible'].append(
+                        html_string)
+                break
+
+        for tabs in soup.findAll(name='oppia-noninteractive-tabs'):
+            # This is wrapped in try catch block since any invalid
+            # content such as int will not be convertible in Beautiful
+            # Soup format.
+            try:
+                tab_content_json = unescape_html(
+                    tabs['tab_contents-with-value'])
+                tab_content_list = json.loads(tab_content_json)
+                for tab_content in tab_content_list:
+                    content_html = tab_content['content']
+                    soup_for_tabs = bs4.BeautifulSoup(
+                        content_html, 'html.parser')
+                    _validate_customization_args_in_soup(
+                        soup_for_tabs, err_dict, html_string)
+            except Exception:
+                if 'oppia-noninteractive-tabs' not in err_dict:
+                    err_dict['oppia-noninteractive-tabs'] = html_string
+                elif html_string not in err_dict[
+                        'oppia-noninteractive-tabs']:
+                    err_dict['oppia-noninteractive-tabs'].append(html_string)
+                break
+
+    return err_dict
+
+
+def _validate_customization_args_in_soup(soup, err_dict, html_string):
+    """Validates customization arguments of Rich Text Components in a soup.
+
+    Args:
+        soup: bs4.BeautifulSoup. The html soup in which customization arguments
+            are to be validated.
+        err_dict: dict. Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+        html_string: str. The html string to be stored in dict if it is invalid.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    # Importing this at the top of the file causes a circular dependency.
```

Comment:
This should be accompanied by an explanation of what the circular dependency is (i.e. what files are involved) and why it cannot be overcome. In general we avoid imports that are not at the top of the file -- it points to problematic structuring of code.

Reply:
Done.